### ISSUE TITLE
Signal regression from 0.9 -> 0.10: only first item emitted on `repeat` source.

### ISSUE THREAD

**Issue by lJoublanc** [functional-streams-for-scala/fs2#907]
The following chunk of code only results in a single item emitted on 0.10.0-M4 (also tested on M2).

```
import fs2._
import cats.effect.IO
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration._

def log[A](prefix : String) : Pipe[IO,A,A] = { in =>
  in.evalMap(i => IO { println(prefix + ">>" + i.toString) ; i} )
}

case class Price(key : Char, price : Double)
object Price {
  def random : IO[Price] = IO {
    val key = 'A' + (scala.math.random() * 10.0).toInt toChar
    val price = scala.math.random()
    Price(key, price)
  }
}

val in : Stream[IO,Price] = (Stream eval Price.random repeat) through log("in")

val sig : Stream[IO, async.immutable.Signal[IO,Option[Price]]] =
  async.immutable.Signal.holdOption(in)

val out : Stream[IO,Option[Price]] =
  sig flatMap (_.discrete) take 20 through log("out")

out.runLog.unsafeRunTimed(1 second)
```

I would expect this to emit the 20 items. It only outputs:
```
in>>Price(H,0.30250069403636715)
out>>None
out>>Some(Price(H,0.30250069403636715))
```

On 0.9.7, with the following code:

```
import fs2._

implicit def strat = Strategy.fromCachedDaemonPool()
def log[A](prefix : String) : Pipe[Task,A,A] =
  _.evalMap(i => Task.delay{ println(prefix + ">>" + i.toString) ; i } )

case class Price(key : Char, price : Double)
object Price {
  def random : Task[Price] = Task {
    val key = 'A' + (scala.math.random() * 10.0).toInt toChar
    val price = scala.math.random()
    Price(key, price)
  }
}

val in : Stream[Task,Price] = (Stream eval Price.random repeat) through log("in")

val sig : Stream[Task, async.immutable.Signal[Task,Option[Price]]] =
  async.immutable.Signal.holdOption(in)

val out : Stream[Task,Option[Price]] =
  sig flatMap (_.discrete) take 20 through log("out")

out.runLog.unsafeRun()
```
This prints out multiple `in`s and the 20 `out`s (can't see the output - it's cut off, sample below):

```
in>>Price(A,0.32971324020042525)
in>>Price(D,0.29919574847430286)
in>>Price(H,0.9111423166252853)
in>>Price(J,0.5500450896250947)
in>>Price(F,0.22712762271018438)
in>>Price(H,0.44739601322878864)
in>>Price(B,0.7649178222004639)
in>>Price(I,0.08453299145449888)
in>>Price(B,0.9298880933223229)
in>>Price(C,0.8484940029373574)
in>>Price(I,0.9747919045403481)
in>>Price(H,0.14794885880903685)
in>>Price(H,0.27389323535519183)
in>>Price(C,0.23817026370678984)
in>>Price(D,0.2903977744468541)
in>>Price(A,0.6209111201757896)
in>>Price(J,0.7897043931941421)
in>>Price(A,0.7167284670123723)
in>>Price(D,0.3363488372388195)
in>>Price(H,0.5082710285336824)
in>>Price(G,0.8791137829005642)
in>>Price(D,0.63527340408363)
out>>Some(Price(H,0.5082710285336824))
in>>Price(G,0.27663035256663004)
in>>Price(I,0.9854773973480855)
in>>Price(F,0.9448985707882575)
in>>Price(A,0.17042652049262041)
out>>Some(Price(F,0.9448985707882575))
in>>Price(C,0.4185048818658893)
out>>Some(Price(A,0.17042652049262041))
out>>Some(Price(C,0.4185048818658893))
in>>Price(I,0.4010571085499287)
out>>Some(Price(I,0.4010571085499287))
in>>Price(E,0.5089171525630363)
in>>Price(J,0.18523549641820947)
Output exceeds cutoff limit.
```

@SystemFw  reproduced this too.

**Comment by SystemFw** [comment]
Minimal repro:

- 0.10 + IO:
```scala
  import fs2._
  import cats.effect._
  import cats.implicits._
  import scala.concurrent.ExecutionContext.Implicits.global

  val in = Stream.range(1,10).covary[IO].observe(Sink.showLinesStdOut[IO, Int])
  
  val out = 
    async.holdOption(in)
      .flatMap(_.discrete)
      .observe(Sink.showLinesStdOut[IO, Option[Int]])

  scala > out.run.unsafeRunSync
  None
  1
  Some(1)
```
- 0.9 + Task
```scala
  import fs2._

  implicit val s = Strategy.fromFixedDaemonPool(4)

  def stdOut[I]: Sink[Task,I] =
    _.map(_.toString).to(_.evalMap(str => Task.delay(Console.out.println(str))))

  val in = Stream.range(1,10).covary[Task].observe(stdOut[Int])

  val out =
    async.holdOption(in)
      .flatMap(_.discrete)
      .observe(stdOut[Option[Int]])
      
scala>  out.run.unsafeRun
 1
 None
 2
 3
 4
 Some(2)
 5
 6
 7
 Some(3) 
 8
 9
  Some(8)
  Some(9)
```

**Comment by AdamChlupacek** [comment]
Soo did a bit of testing

```
  Stream.eval(async.signalOf[IO, Option[Price]](None)).flatMap{ s =>
    
    Stream(s) merge
    in.flatMap( i => Stream.eval_(s.set(Some(i)))).drain


  }.flatMap(_.discrete.take(10) through log("out")).run.unsafeRunSync()
```

Does not work, while this works: 

```
  Stream.eval(async.signalOf[IO, Option[Price]](None)).flatMap{ s =>

    in.flatMap( i => Stream.eval_(s.set(Some(i)))).drain merge
    s.discrete.take(10) through log("out")

  }.run.unsafeRunSync()
```

From that i would assume that it may have to do something with the singleton stream optimisation in flatMap? just a blind shot but only thing I could think of.

**Comment by AdamChlupacek** [comment]
I take it back for it to have something to do with the flatMap, although seems like it is an issue with calling discrete in the flatMap, when I swapped it for `Ref` it did work, also I have noticed that in the case of the streams getting stuck, the `race` in merge finishes only once, which I assumed that it should finish twice with left and then start the echo on the right side

**Comment by mpilquist** [comment]
I think this is working as intended. The expression `s flatMap f` means "lazily map f over each element and concatenate the results". Let's look at the minimization:

```
val in = Stream.range(1,10).covary[IO].observe(Sink.showLinesStdOut[IO, Int])
  
val out = 
  async.holdOption(in)
    .flatMap(_.discrete)
    .observe(Sink.showLinesStdOut[IO, Option[Int]])
```

Here, `async.holdOption(in)` returns a singleton stream consisting of a single element of type `Signal`. We pull the single element from it and then return a stream created via `signal.discrete`. The stream interpreter pulls from the discrete stream and doesn't return to the `async.holdOption(in)` stream until it reaches the end of the discrete stream. Hence, the overall stream stalls.

To fix, I think we should change `hold` and `holdOption` so that they take a `discrete: Boolean` argument and return a `Stream[F,A]`. Rather than emitting the signal, the implementation would merge the source with either a discrete stream or continuous stream. I'll open a PR shortly.

**Comment by SystemFw** [comment]
What I find strange is the difference in behaviour between 0.10 and 0.9. How come the 0.9 version does indeed proceed in observing the changes?

**Comment by mpilquist** [comment]
@SystemFw I'm not sure - the 0.10 interpreter is much easier to reason about. The 0.9 interpreter is extremely hard to reason about.

**Comment by SystemFw** [comment]
@mpilquist Fair enough :) Is the interpreter code in `Algebra` ?

I also have another question, I would expect this (I moved observe):
```scala
val in = Stream.range(1,10).covary[IO].observe(Sink.showLinesStdOut[IO, Int])
  
val out = 
  async.holdOption(in)
    .flatMap(_.discrete.observe(Sink.showLinesStdOut[IO, Option[Int]]))    
```
to print the changes (even though it would probably still stall), but it does not...

**Comment by mpilquist** [comment]
@SystemFw That should still result in a single element only being observed, once from the observation of `in` and once from the observation of `discrete`.

**Comment by SystemFw** [comment]
Yeah, I think I am overestimating the power of `merge`: even though it runs things concurrently, you still need to pull to evaluate the next element, and you can't pull from the Stream returned by `hold` until `discrete` is done (since you're flatMapping), but otoh `discrete` can only advance if you pull again...hence the stall.

**Comment by pchlupacek** [comment]
@mpilquist @SystemFw 
My observation from porting fs2-http to 0.10M3 was that 0.10 stack is so much faster than some tests start to behave differently, if they were not perfectly defined. I don't believe however this is that case though. 

I understand issue here is that 
 ```scala
val in = Stream.range(1,10).covary[IO].observe(Sink.showLinesStdOut[IO, Int])
```
will give you 10 values and then terminates. However there is no guarantee that discrete stream as output form the hold combinator will contain them all. However it must always contain at least the last element (9) in our case. 

I think the original issue is not what we are really solving there, because to my understanding the input was infinite stream of values, so `in` would never terminate, yet the `flatMap { _.discrete } evalMap { outputToConsole } did not produce 20 output elements (Unit). I think we must have something odd with 0.10 interpreter if  in this case it won't behave as 0.9 interpreter. 

I am not sure If I was following your explanation correctly @mpilquist, but let's remove the observe, and replace it with evalMap to make reasoning clearer, and then I am quite confident that this program: 

```scala
scala> val in:Stream[IO, Int] = Stream.range(0, Int.MaxValue).covary[IO]
in: fs2.Stream[cats.effect.IO,Int] = Stream(..)
scala> def log(in: Int): IO[Unit] = IO.suspend { IO.pure { println(s">>>> $in") } }
log: (in: Int)cats.effect.IO[Unit]
scala> val signal = async.hold(-1, in)
signal: fs2.Stream[cats.effect.IO,fs2.async.immutable.Signal[cats.effect.IO,Int]] = Stream(..)
scala> (signal.flatMap { _.discrete } evalMap { log } take 20) run
res2: cats.effect.IO[Unit] = IO$858784817

scala> res2.unsafeRunTimed(5.seconds)
>>>> 0
res5: Option[Unit] = None

scala> res2.unsafeRunTimed(5.seconds)
>>>> 0
>>>> 1
res6: Option[Unit] = None

scala> res2.unsafeRunTimed(5.seconds)
>>>> 1
res7: Option[Unit] = None

scala> res2.unsafeRunTimed(5.seconds)
>>>> 0
res8: Option[Unit] = None

```

should produce exactly 20 print-lines likely  NOT consecutive int values. Which, sadly does not in 0.10 M4. In fact we are using this construct so often in our server code, that I can't imagine the 0.9 and 0.10 shall be different there.

**Comment by pchlupacek** [comment]
@mpilquist after discussing this with @AdamChlupacek I think to understand your point which comes from different behaviour of merge since 0.10 : 

the hold is defined like 

```scala 

def hold[F[_],A](initial: A, source:Stream[F,A])(implicit F: Effect[F], ec: ExecutionContext): Stream[F,immutable.Signal[F,A]] =
    Stream.eval(fs2.async.signalOf[F,A](initial)) flatMap { sig =>
      Stream(sig).merge(source.flatMap(a => Stream.eval_(sig.set(a))))
    }

```

now in 0.9 the merge was eager, in sense that both sides of the stream were running until each of them produced al least one nonempty chunk, with winning side emitting first to downstream. 

In 0.10 however the `drained` stream "cares" about empty Segment and thus stoping evaluation of the drained  stream once that empty segment is produced by It. 

That really causes the `feeder` stream to be evaluated only once, or twice if time allows. 

Although I think this is easily fixable by changing hold implementation a bit, I think this is major _invisible_ behaviour change that is worth of discussing, specifically the impact of it (rendering quite few programs to be impacted). I am not biased now whether the new behaviour is correct but there are lot of usages of idioms using `mergeDrainR`, `mergeDrainL` for exactly that purpose, i.e. running some stream in background that will be teared down when final stream is terminated. 

@mpilquist WDYT? I think is perhaps worth of discussing and agreeing on this in another issue and then resolving this accordingly ?

**Comment by SystemFw** [comment]
@pchlupacek That's an interesting point about `merge`
